crates/exec-wasmtime/src/runtime/net/mod.rs

```diff
@@ -0,0 +1,108 @@
+    FileCaps::FILESTAT_GET | FileCaps::FDSTAT_SET_FLAGS | FileCaps::POLL_READWRITE | FileCaps::READ
+});
+
+static CONNECT_CAPS: Lazy<FileCaps> = Lazy::new(|| {
+    FileCaps::FILESTAT_GET
+        | FileCaps::FDSTAT_SET_FLAGS
+        | FileCaps::POLL_READWRITE
+        | FileCaps::READ
+        | FileCaps::WRITE
+});
+
+pub fn listen_file(
+    file: &ListenFile,
+    certs: Vec<Certificate>,
+    key: &Zeroizing<Vec<u8>>,
+) -> Result<(Box<dyn WasiFile>, FileCaps)> {
+    let (addr, port) = match file {
+        ListenFile::Tcp { addr, port, .. } | ListenFile::Tls { addr, port, .. } => (addr, port),
+    };
+    let tcp = std::net::TcpListener::bind((addr.as_str(), *port))?;
+    let tcp = TcpListener::from_std(tcp);
+    let file = match file {
+        ListenFile::Tcp { .. } => wasmtime_wasi::net::Socket::from(tcp).into(),
+        ListenFile::Tls { .. } => {
+            let cfg = rustls::ServerConfig::builder()
+                .with_cipher_suites(DEFAULT_TLS_CIPHER_SUITES.deref())
+                .with_kx_groups(DEFAULT_TLS_KX_GROUPS.deref())
+                .with_protocol_versions(DEFAULT_TLS_PROTOCOL_VERSIONS.deref())?
+                .with_no_client_auth() // TODO: https://github.com/enarx/enarx/issues/1547
+                .with_single_cert(certs, PrivateKey(key.deref().clone()))?;
+            tls::Listener::new(tcp, Arc::new(cfg)).into()
+        }
+    };
+    Ok((file, *LISTEN_CAPS))
+}
+
+pub fn connect_file(
+    file: &ConnectFile,
+    certs: Vec<Certificate>,
+    key: &Zeroizing<Vec<u8>>,
+) -> Result<(Box<dyn WasiFile>, FileCaps)> {
+    let (host, port) = match &file {
+        ConnectFile::Tcp { host, port, .. } | ConnectFile::Tls { host, port, .. } => (host, port),
+    };
+    let tcp = std::net::TcpStream::connect((host.as_str(), *port))?;
+    let tcp = TcpStream::from_std(tcp);
+    let file = match file {
+        ConnectFile::Tcp { .. } => wasmtime_wasi::net::Socket::from(tcp).into(),
+        ConnectFile::Tls { .. } => {
+            let mut server_roots = RootCertStore::empty();
+            server_roots.add_server_trust_anchors(webpki_roots::TLS_SERVER_ROOTS.0.iter().map(
+                |ta| {
+                    rustls::OwnedTrustAnchor::from_subject_spki_name_constraints(
+                        ta.subject,
+                        ta.spki,
+                        ta.name_constraints,
+                    )
+                },
+            ));
+            let cfg = rustls::ClientConfig::builder()
+                .with_cipher_suites(DEFAULT_TLS_CIPHER_SUITES.deref())
+                .with_kx_groups(DEFAULT_TLS_KX_GROUPS.deref())
+                .with_protocol_versions(DEFAULT_TLS_PROTOCOL_VERSIONS.deref())?
+                .with_root_certificates(server_roots)
+                .with_single_cert(certs, PrivateKey(key.deref().clone()))?;
+
+            tls::Stream::connect(tcp, host, Arc::new(cfg))?.into()
+        }
+    };
+    Ok((file, *CONNECT_CAPS))
+}
```

Comment:
So, I would generally approve.. but maybe the introduction of this file is worth its own commit.

Reply:
It's from
https://github.com/enarx/enarx/blob/de9304547d0ca7a1e947533ed0ef0d2fa9f2f8f5/crates/exec-wasmtime/src/loader/compiled/mod.rs#L50-L86
and
https://github.com/enarx/enarx/blob/de9304547d0ca7a1e947533ed0ef0d2fa9f2f8f5/crates/exec-wasmtime/src/loader/requested.rs#L285-L319

Originally, this was the catalyst for me to start doing this, because I had to add functionality to add custom roots per-connection, so I had to create a TLS config per-connection, I was working on this on the plane.

Originally I had several commits here, which I may even be able to find in the reflog, but in my opinion commits should be "atomic", should compile and pass tests. To achieve that I would have to come up with another "intermediate" structure first and then do this change, if you think that's worth the time and effort spent, then I can do that.

Reply:
Additionally, this file will change significantly in one of follow-up PRs with the helper functions removed